util/pingpong.c

```diff
@@ -0,0 +1,2111 @@
+
+	return 0;
+}
+
+int ft_open_fabric_res(struct ct_pingpong *ct)
+{
+	int ret;
+
+	ret = fi_fabric(ct->fi->fabric_attr, &(ct->fabric), NULL);
+	if (ret) {
+		FT_PRINTERR("fi_fabric", ret);
+		return ret;
+	}
+
+	ret = fi_eq_open(ct->fabric, &(ct->eq_attr), &(ct->eq), NULL);
+	if (ret) {
+		FT_PRINTERR("fi_eq_open", ret);
+		return ret;
+	}
+
+	ret = fi_domain(ct->fabric, ct->fi, &(ct->domain), NULL);
+	if (ret) {
+		FT_PRINTERR("fi_domain", ret);
+		return ret;
+	}
+
+	return 0;
+}
+
+int ft_alloc_active_res(struct ct_pingpong *ct, struct fi_info *fi)
+{
+	int ret;
+
+	ret = ft_alloc_msgs(ct);
+	if (ret)
+		return ret;
+
+	if (ct->cq_attr.format == FI_CQ_FORMAT_UNSPEC) {
+		ct->cq_attr.format = FI_CQ_FORMAT_CONTEXT;
+	}
+
+	ct->cq_attr.wait_obj = FI_WAIT_NONE;
+	ct->cq_attr.size = fi->tx_attr->size;
+	ret = fi_cq_open(ct->domain, &(ct->cq_attr), &(ct->txcq), &(ct->txcq));
+	if (ret) {
+		FT_PRINTERR("fi_cq_open", ret);
+		return ret;
+	}
+
+	ct->cq_attr.wait_obj = FI_WAIT_NONE;
+	ct->cq_attr.size = fi->rx_attr->size;
+	ret = fi_cq_open(ct->domain, &(ct->cq_attr), &(ct->rxcq), &(ct->rxcq));
+	if (ret) {
+		FT_PRINTERR("fi_cq_open", ret);
+		return ret;
+	}
+
+	if (fi->ep_attr->type == FI_EP_RDM || fi->ep_attr->type == FI_EP_DGRAM) {
+		if (fi->domain_attr->av_type != FI_AV_UNSPEC)
+			ct->av_attr.type = fi->domain_attr->av_type;
+
+		ret = fi_av_open(ct->domain, &(ct->av_attr), &(ct->av), NULL);
+		if (ret) {
+			FT_PRINTERR("fi_av_open", ret);
+			return ret;
+		}
+	}
+
+	ret = fi_endpoint(ct->domain, fi, &(ct->ep), NULL);
+	if (ret) {
+		FT_PRINTERR("fi_endpoint", ret);
+		return ret;
+	}
+
+	return 0;
+}
+
+static int dupaddr(void **dst_addr, size_t *dst_addrlen,
+		void *src_addr, size_t src_addrlen)
+{
+	*dst_addr = malloc(src_addrlen);
+	if (!*dst_addr) {
+		FT_ERR("address allocation failed");
+		return EAI_MEMORY;
+	}
+	*dst_addrlen = src_addrlen;
+	memcpy(*dst_addr, src_addr, src_addrlen);
+	return 0;
+}
+
+static int getaddr(char *node, char *service,
+			struct fi_info *hints, uint64_t flags)
+{
+	int ret;
+	struct fi_info *fi;
+
+	if (!node && !service) {
+		if (flags & FI_SOURCE) {
+			hints->src_addr = NULL;
+			hints->src_addrlen = 0;
+		} else {
+			hints->dest_addr = NULL;
+			hints->dest_addrlen = 0;
+		}
+		return 0;
+	}
+
+	ret = fi_getinfo(FT_FIVERSION, node, service, flags, hints, &fi);
+	if (ret) {
+		FT_PRINTERR("fi_getinfo", ret);
+		return ret;
+	}
+	hints->addr_format = fi->addr_format;
+
+	if (flags & FI_SOURCE) {
+		ret = dupaddr(&hints->src_addr, &hints->src_addrlen,
+				fi->src_addr, fi->src_addrlen);
+	} else {
+		ret = dupaddr(&hints->dest_addr, &hints->dest_addrlen,
+				fi->dest_addr, fi->dest_addrlen);
+	}
+
+	fi_freeinfo(fi);
+	return ret;
+}
+
+void eq_readerr(struct fid_eq *eq, const char *eq_str)
+{
+	struct fi_eq_err_entry eq_err;
+	int rd;
+
+	rd = fi_eq_readerr(eq, &eq_err, 0);
+	if (rd != sizeof(eq_err)) {
+		FT_PRINTERR("fi_eq_readerr", rd);
+	} else {
+		FT_EQ_ERR(eq, eq_err, NULL, 0);
+	}
+}
+
+
```

Comment:
Nit pick: either have 1 blank line between functions or 2 blank lines between functions -- be consistent.  😄